File: src/main/java/link/infra/packwiz/installer/ui/CLIHandler.java
package link.infra.packwiz.installer.ui;

import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.Future;

public class CLIHandler implements IUserInterface {

	@Override
	public void handleException(Exception e) {
		e.printStackTrace();
	}

	@Override
	public void show(InputStateHandler h) {}

	@Override
	public void submitProgress(InstallProgress progress) {
		StringBuilder sb = new StringBuilder();
		if (progress.hasProgress) {
			sb.append('(');
			sb.append(progress.progress);
			sb.append('/');
			sb.append(progress.progressTotal);
			sb.append(") ");
		}
		sb.append(progress.message);
		System.out.println(sb.toString());
	}

	@Override
	public void executeManager(Runnable task) {
		task.run();
		System.out.println("Finished successfully!");
	}

	@Override
	public Future<Boolean> showOptions(List<IOptionDetails> options) {
		for (IOptionDetails opt : options) {
			opt.setOptionValue(true);
			System.out.println("Warning: accepting option " + opt.getName() + " as option choosing is not implemented in the CLI");
		}
		CompletableFuture<Boolean> future = new CompletableFuture<>();
		future.complete(false); // Can't be cancelled!
		return future;
	}

	@Override
	public Future<IExceptionDetails.ExceptionListResult> showExceptions(List<IExceptionDetails> opts, int numTotal, boolean allowsIgnore) {
		CompletableFuture<IExceptionDetails.ExceptionListResult> future = new CompletableFuture<>();
		future.complete(IExceptionDetails.ExceptionListResult.CANCEL);
		return future;
	}

}

File: src/main/kotlin/link/infra/packwiz/installer/ui/CLIHandler.kt
package link.infra.packwiz.installer.ui

import link.infra.packwiz.installer.ui.IUserInterface.ExceptionListResult
import java.util.concurrent.CompletableFuture
import java.util.concurrent.Future

class CLIHandler : IUserInterface {
	override fun handleException(e: Exception) {
		e.printStackTrace()
	}

	override fun show(handler: InputStateHandler) {}
	override fun submitProgress(progress: InstallProgress) {
		val sb = StringBuilder()
		if (progress.hasProgress) {
			sb.append('(')
			sb.append(progress.progress)
			sb.append('/')
			sb.append(progress.progressTotal)
			sb.append(") ")
		}
		sb.append(progress.message)
		println(sb.toString())
	}

	override fun executeManager(task: () -> Unit) {
		task()
		println("Finished successfully!")
	}

	override fun showOptions(options: List<IOptionDetails>): Future<Boolean> {
		for (opt in options) {
			opt.optionValue = true
			// TODO: implement option choice in the CLI?
			println("Warning: accepting option " + opt.name + " as option choosing is not implemented in the CLI")
		}
		return CompletableFuture<Boolean>().apply {
			complete(false) // Can't be cancelled!
		}
	}

	override fun showExceptions(exceptions: List<ExceptionDetails>, numTotal: Int, allowsIgnore: Boolean): Future<ExceptionListResult> {
		val future = CompletableFuture<ExceptionListResult>()
		future.complete(ExceptionListResult.CANCEL)
		return future
	}
}